Implement setter function to check MC signal from mother to daughter

feisenhu opened the pull request:
Implemented a setter function in the MCprong files to give the option, when defining a MCsignal in the MCSignalLibrary.h to check the MC signal from the mother to the daughter (In time).

Therefore the MC signal needs the mother pdg code as first entry and its daughter as second:
e.g. 
```
  if (!nameStr.compare("LFQdecayToE")) {
    MCProng prong(2, {900, 11}, {true, true}, {false, false}, {0, 0}, {0, 0}, {false, false});
    prong.SetSignalInTime(0, true);                                                    // set direction to check for daughters (true, in time) or for mothers (false, back in time)
    signal = new MCSignal(name, "LF meson  + quarkonia decays into e", {prong}, {-1}); // pi0,eta,eta',rho,omega,phi,jpsi,psi2s mesons
    return signal;
  }
```

The CheckProng function in MCSignal.h then checks the boolean `fCheckGenerationsInTime` whether we have to check in the next generation for a mother or daughters.

If we look for daughters, we check if there is one, and if so if it has a pdg code of the next generation.
If it is true it uses the first daughter that fulfils that requirement and sets it as current particle to check for further generations.

(Note: There is no option currently to check multiple daughters in the same generation that would fulfil the MC signal.)

alibuild commented:
Error while checking build/O2Physics/o2 for 7f828ff81d2b30704db1d6158cc7c8dc601cb020 at 2022-07-04 11:44:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.h:102:21: error: 'MCProng::fUseANDonSourceBitMap' will be initialized after [-Werror=reorder]
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.h:96:8: error:   'bool MCProng::fCheckGenerationsInTime' [-Werror=reorder]
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.cxx:52:1: error:   when initialized here [-Werror=reorder]
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/945/7f828ff81d2b30704db1d6158cc7c8dc601cb020/build_O2Physics_o2/pretty.html).

alibuild commented:
Error while checking build/O2Physics/o2 for a31fd4f287efeb5df2da3734283563d5cdc8fe80 at 2022-07-04 14:14:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.h:102:21: error: 'MCProng::fUseANDonSourceBitMap' will be initialized after [-Werror=reorder]
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.h:96:8: error:   'bool MCProng::fCheckGenerationsInTime' [-Werror=reorder]
/sw/SOURCES/O2Physics/945-slc7_x86-64/0/PWGDQ/Core/MCProng.cxx:52:1: error:   when initialized here [-Werror=reorder]
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/945/a31fd4f287efeb5df2da3734283563d5cdc8fe80/build_O2Physics_o2/pretty.html).

iarsene commented on PWGDQ/Core/MCProng.h:
```diff
@@ -79,14 +79,15 @@ class MCProng
   void SetSources(int generation, uint64_t bits, uint64_t exclude = 0, bool useANDonSourceBits = true);
   void SetSourceBit(int generation, int sourceBit, bool exclude = false);
   void SetUseANDonSourceBits(int generation, bool option = true);
+  void SetSignalInTime(int generation, bool intime = false); // set variable to check generations in time or back in time (default)
```
This function does not need the first parameter (generation), since the rule to go back in time or in time, holds for the entire prong. So please remove it, only the "intime" parameter is used

iarsene commented on PWGDQ/Core/MCProng.cxx:
```diff
@@ -100,13 +103,22 @@ void MCProng::SetUseANDonSourceBits(int generation, bool option /*=true*/)
 }
 
+//________________________________________________________________________________________________________________
+void MCProng::SetSignalInTime(int generation, bool intime /*=false*/)
```
The "generation" parameter is not needed since the "intime" holds for the whole prong history

iarsene commented on PWGDQ/Core/MCProng.cxx:
```diff
@@ -100,13 +103,22 @@ void MCProng::SetUseANDonSourceBits(int generation, bool option /*=true*/)
+//________________________________________________________________________________________________________________
+void MCProng::SetSignalInTime(int generation, bool intime /*=false*/)
+{
+  if (generation < 0 || generation >= fNGenerations) {
+    return;
+  }
```
these lines can be removed

iarsene commented on PWGDQ/Core/MCSignal.h:
```diff
@@ -147,17 +147,36 @@ bool MCSignal::CheckProng(int i, bool checkSources, const U& mcStack, const T& t
+      if (currentMCParticle.has_daughters() && j < fProngs[i].fNGenerations - 1) {
+        const auto& daughtersSlice = currentMCParticle.template daughters_as<U>();
+        for (auto& d : daughtersSlice) {
+          if (abs(d.pdgCode()) == fProngs[i].fPDGcodes[j + 1]) {
+            currentMCParticle = d;
+            break;
```
Here you check the absolute of the PDG code, however, the user has the option to choose between checking both charges (using the abs of the PDG code), or not. So the latter situation is overwritten here. Could you fix it? Ideally one should use here the MCProng::TestPDG() since that will take into account automatically these options.

iarsene commented on PWGDQ/Core/MCSignal.h:
```diff
@@ -202,23 +221,47 @@ bool MCSignal::CheckProng(int i, bool checkSources, const U& mcStack, const T& t
+        if (currentMCParticle.has_daughters() && j < fProngs[i].fNGenerations - 1) {
+          const auto& daughtersSlice = currentMCParticle.template daughters_as<U>();
+          for (auto& d : daughtersSlice) {
+            if (abs(d.pdgCode()) == fProngs[i].fPDGcodes[j + 1]) {
+              currentMCParticle = d;
+              break;
```
same comment as above

feisenhu commented:
@iarsene 
Tank you for your feedback.
I have included your comments:
- removed the `generation` parameter in the SetSignalInTime() function
- added a comment for the else {} case, to point out that branch is changing to daughter prong which fulfils the pdg requirement first
- using the TestPDG() function to check if the daughter PDG is in the next generation
   I now use `fProngs[i].TestPDG(j + 1, d.pdgCode())` to check that daughter has correct pdg code.
   
In case there is something else that I should change please let me know.
kind regards

iarsene commented on PWGDQ/Core/MCSignal.h:
```diff
@@ -202,23 +221,48 @@ bool MCSignal::CheckProng(int i, bool checkSources, const U& mcStack, const T& t
+        if (!currentMCParticle.has_daughters() && j < fProngs[i].fNGenerations - 1) {
+          return false;
+        }
+        if (currentMCParticle.has_daughters() && j < fProngs[i].fNGenerations - 1) {
+          // currentMCParticle = currentMCParticle.template mothers_first_as<U>();
+          // currentMCParticle = mcStack.iteratorAt(currentMCParticle.mothersIds()[0]);
+        }
```
This if statement doesn't do anything, should be removed

feisenhu replied:
You are right. Thank you for pointing out. :)
I have removed these lines.

iarsene commented:
Hi @feisenhu,
Thanks for the changes. There is just a small issue i still commented on, otherwise I am fine with the changes.

feisenhu commented:
> Hi @feisenhu,
Thanks for the changes. There is just a small issue i still commented on, otherwise I am fine with the changes.

You are right. Thank you for pointing out. :)
I have removed these lines.
